plugin/caffe/caffe_data_iter.cc

```diff
@@ -0,0 +1,264 @@
+    CHECK_EQ(out_.batch_size, batch_size_) << "Internal Error: batch size mismatch";
+
+    if (loc_ + batch_size_ <= param_.num_examples) {
+      batch_data_.dptr_ = top_[DATA]->mutable_cpu_data();
+      batch_label_.dptr_ = top_[LABEL]->mutable_cpu_data();
+
+      out_.data.clear();
+      if (param_.flat) {
+        out_.data.push_back(batch_data_.FlatTo2D<cpu, Dtype>());
+      } else {
+        out_.data.push_back(batch_data_);
+      }
+      out_.data.push_back(batch_label_);
+      loc_ += batch_size_;
+      return true;
+    }
+
+    return false;
+  }
+
+  virtual const TBlobBatch &Value(void) const {
+    return out_;
+  }
+
+ private:
+  /*! \brief indexes into top_ */
+  enum { DATA = 0, LABEL, NR_SUPPORTED_TOP_ITEMS };
+
+  /*! \brief MNISTCass iter params */
+  CaffeDataParam param_;
+  /*! \brief Shape scalar values */
+  index_t batch_size_, channels_, width_, height_;
+  /*! \brief Caffe data layer */
+  boost::shared_ptr<caffe::Layer<Dtype> >  caffe_data_layer_;
+  /*! \brief batch data blob */
+  mxnet::TBlob batch_data_;
+  /*! \brief batch label blob */
+  mxnet::TBlob batch_label_;
+  /*! \brief Output blob data for this iteration */
+  TBlobBatch out_;
+  /*! \brief Bottom and top connection-point blob data */
+  std::vector<::caffe::Blob<Dtype>*> bottom_, top_;
+  /*! \brief Cleanup these blobs on exit */
+  std::list<std::unique_ptr<::caffe::Blob<Dtype>>> cleanup_blobs_;
+  /*! \brief Blobs done so far */
+  std::atomic<size_t>  loc_;
+};  // class CaffeDataIter
+
+class CaffeDataIterWrapper : public IIterator<DataBatch>
+{
+ public:
+    virtual void Init(const std::vector<std::pair<std::string, std::string> >& kwargs) {
+      param_.InitAllowUnknown(kwargs);
+      switch (param_.dtype) {
+        case mshadow::kFloat32:
+          this->loader_.reset(new CaffeDataIter<float>());
+          break;
+        case mshadow::kFloat64:
+          this->loader_.reset(new CaffeDataIter<double>());
+          break;
+        case mshadow::kFloat16:
+          LOG(FATAL) << "float16 layer is not supported by caffe";
+          return;
+        default:
+          LOG(FATAL) << "Unsupported type " << param_.dtype;
+          return;
+      }
+      if(loader_) {
+        loader_->Init(kwargs);
+      }
+    }
+    virtual void BeforeFirst(void) {
+      CHECK_NOTNULL(loader_.get());
+      return loader_->BeforeFirst();
+    }
+    virtual bool Next(void) {
+      CHECK_NOTNULL(loader_.get());
+
+      if (!loader_->Next()) return false;
+      const TBlobBatch& batch = loader_->Value();
+      if (out_ == nullptr) {
+        // allocate databatch
+        out_.reset(new DataBatch());
+        out_->num_batch_padd = batch.num_batch_padd;
+        out_->data.resize(batch.data.size());
+        out_->index.resize(batch.batch_size);
+        for (size_t i = 0; i < batch.data.size(); ++i) {
+          out_->data.at(i) = NDArray(batch.data[i].shape_, Context::CPU());
```

Comment:
putting type flag in PrefetcherIter and changing PrefetcherIter to base, with prefetch size override as we discussed offline